official/boosted_trees/train_higgs.py

```diff
@@ -0,0 +1,184 @@
+So you need to install recent enough version of Tensorflow to use this example.
+
+The training data is by default the first million examples out of 11M examples,
+and eval data is by default the last million examples.
+They are controlled by --train_start, --train_count, --eval_start, --eval_count.
+e.g. to train over the first 10 million examples instead of 1 million:
+$ python train_higgs.py --n_trees=100 --max_depth=6 --learning_rate=0.1 \
+    --model_dir=/tmp/higgs_model --train_count=10000000
+
+Training history and metrics can be inspected using tensorboard.
+$ tensorboard --logdir=/tmp/higgs_model
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import sys
+
+import numpy as np
+import tensorflow as tf
+
+NPZ_FILE = 'HIGGS.csv.gz.npz'  # numpy compressed file containing 'data' array
+
+
+def parse_args():
+  """Parses arguments and returns a tuple (known_args, unparsed_args)."""
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--data_dir', type=str, default='/tmp/higgs_data',
+      help='Path to directory containing the higgs dataset.')
+  parser.add_argument(
+      '--model_dir', type=str, default='/tmp/higgs_model',
+      help='The directory where the model will be stored.')
+  parser.add_argument(
+      '--train_start', type=int, default=0,
+      help='Start index of train examples within the data.')
+  parser.add_argument(
+      '--train_count', type=int, default=1000000,
+      help='Number of train examples within the data.')
+  parser.add_argument(
+      '--eval_start', type=int, default=10000000,
+      help='Start index of eval examples within the data.')
+  parser.add_argument(
+      '--eval_count', type=int, default=1000000,
+      help='Number of eval examples within the data.')
+
+  parser.add_argument(
+      '--n_trees', type=int, default=100, help='Number of trees to build.')
+  parser.add_argument(
+      '--max_depth', type=int, default=6, help='Maximum depths of each tree.')
+  parser.add_argument(
+      '--learning_rate', type=float, default=0.1,
+      help='Maximum depths of each tree.')
+  return parser.parse_known_args()
+
+
+def read_higgs_data():
+  """Reads higgs data from csv and returns train and eval data."""
+  npz_filename = os.path.join(FLAGS.data_dir, NPZ_FILE)
+  try:
+    # gfile allows numpy to read data from network data sources as well.
+    with tf.gfile.Open(npz_filename) as npz_file:
+      with np.load(npz_file) as npz:
+        data = npz['data']
+  except Exception as e:
+    raise RuntimeError(
+        'Error loading data; use data_download.py to prepare the data:\n{}: {}'
+        .format(type(e).__name__, e))
+  return (data[FLAGS.train_start:FLAGS.train_start+FLAGS.train_count],
+          data[FLAGS.eval_start:FLAGS.eval_start+FLAGS.eval_count])
+
+
+def make_inputs_from_np_arrays(features_np, label_np):
```

Comment:
Could you comment on why so much of the data pipeline is done outside of tf.data.Dataset? I vaguely recall there being some nuances around boosted trees and datasets, but I don't recall the details.

Reply:
There are several reasons; mostly to utilize benefits of numpy arrays.
1) bucketized_column requires static bucket boundaries, which are generated from numpy arrays in this example.
2) data.Dataset is streaming data and could be slow. The focus of this example is to train a model with data in memory. (data streaming use case might be a separate example.)
3) we've seen examples of gradient boosted trees algorithms use various data formats (like svm text formats or many custom formats), which Tensorflow doesn't support, and not likely support any time soon. OTOH, there are many libraries like pandas or sklearn that support various data sources and eventually return numpy matrices . So it's worth to show a data pipeline that starts with numpy array, not from raw files.

Reply:
Just for the reference, 
- the data_download.py tool downloads the csv.gz file and parses using pandas to save into a numpy compressed file.
- numpy save/load is pretty light weight - both in several seconds for the float matrix of size (10 million * 29).
- pandas.read_csv() takes 2~3 mins to create ndarray from the csv.gz file in 6-core workstation (bogomips 7000 per core).
- numpy.loadtext() or genfromtxt() takes ~10 mins.
- tf.data pipeline to read through and parse the entire csv file takes ~ 9 mins with TextLineDataset 
 and map.

The numpy file can be generated once and reused multiple times without overhead. 
Since most users would run training many times with different parameters,
we wanted to reduce the total cost (and better user experience).